src/core/api.c

```diff
@@ -1321,14 +1321,12 @@ MsQuicStreamReceiveComplete(
         Stream,
         BufferLength);
 
-    CxPlatRwLockAcquireShared(&Stream->ReceiveCompleteInlineLock);
-    BOOLEAN ReceiveCompleteInline = Stream->Flags.ReceiveCallActive;
-    InterlockedExchangeAdd64(
-        (int64_t*)&Stream->RecvCompletionLength, (int64_t)BufferLength);
-    CxPlatRwLockReleaseShared(&Stream->ReceiveCompleteInlineLock);
-
-    if (ReceiveCompleteInline) {
-        goto Exit; // No need to queue a completion operation when run inline
+    uint64_t RecvCompletionLength = InterlockedExchangeAdd64(
```

Comment:
If the app passes a really big `BufferLength` in, it could mess up this logic. We should probably have some defense against that, maybe limiting `BufferLength <= (UINT64_MAX >> 2)`?

Reply:
The logic was already vulnerable to an overflow, we only reach that condition earlier now (but still for a value that should be unreachable in a valid scenario).

`BufferLength <= (UINT64_MAX >> 2)` would not be enough to prevent it, because `Stream->RecvCompletionLength` might not be null.

If we care about being safe, I think we can use the second highest bit of `Stream->RecvCompletionLength` as a canary:
- we implement the check you suggest
- after an addition, we check the new value and if the bit is set, we abort

Reply:
I've added the check of overflow. Is this your intended?

Reply:
Generally, yes. Thanks!